nutrition_labels/grant_tagger.py

```diff
@@ -11,248 +39,358 @@
-        options are 'bert' and 'scibert', this doesn't need to be defined if you\
-        aren't using bert",
-        default=None
-    )
+    training_data = pd.read_csv(config["data"]["training_data_file"])
 
-    return parser
+    if "grants_text_data_file" in config["data"]:
+        # Merge training data with another dataset containing the
+        # text columns for predicting on
+        grants_data = pd.read_csv(config["data"]["grants_text_data_file"])
+        right_id = config["data"]["grants_text_data_file_id"]
+        training_data = pd.merge(
+            training_data,
+            grants_data.drop_duplicates(subset="Internal ID")[
+                prediction_cols + [right_id]
+            ],
+            how="left",
+            left_on=train_data_id,
+            right_on=right_id,
+        )
 
-if __name__ == '__main__':
+    return training_data
 
-    parser = create_argparser()
-    args = parser.parse_args()
 
-    datestamp = datetime.now().date().strftime('%y%m%d')
+def train_several_models(config):
 
-    data = pd.read_csv(args.training_data_file)
+    # Load data and parameters
+    vectorizer_types = ast.literal_eval(config["models"]["vectorizer_types"])
+    classifier_types = ast.literal_eval(config["models"]["classifier_types"])
 
-    grant_tagger = GrantTagger(
-        ngram_range=(1, 2),
-        test_size=0.25,
-        vectorizer_type=args.vectorizer_type,
-        model_type=args.model_type,
-        bert_type=args.bert_type,
-        relevant_sample_ratio=args.relevant_sample_ratio
+    relevant_sample_ratio = config.getfloat("params", "relevant_sample_ratio")
+    test_size = config.getfloat("params", "test_size")
+    split_seed = config["params"].getint("split_seed")
+
+    prediction_cols = ast.literal_eval(config["data"]["prediction_cols"])
+    label_name = config["data"]["label_name"]
+    train_data_id = config["data"]["training_data_file_id"]
+    training_data = load_training_data(config, prediction_cols, train_data_id)
+
+    config_version = "".join(config["DEFAULT"]["version"].split("."))[2:]
+
+    # Train and save several models
+    for vectorizer_type in vectorizer_types:
+        for classifier_type in classifier_types:
+            print(f"Training for {vectorizer_type} + {classifier_type} ...")
+
+            if not split_seed:
+                split_seed = np.random.randint(100)  # Pick randomly if not given
+            grant_tagger = GrantTagger(
+                test_size=test_size,
+                relevant_sample_ratio=relevant_sample_ratio,
+                split_seed=split_seed,
+                vectorizer_type=vectorizer_type,
+                classifier_type=classifier_type,
+                prediction_cols=prediction_cols,
+                label_name=label_name,
+            )
+            X_vect, y = grant_tagger.transform(training_data, train_data_id)
+
+            X_train, X_test, y_train, y_test = grant_tagger.split_data(X_vect, y)
```

Comment:
as discussed previously I would split data before as a preprocessing step and combined with the comment about the pipeline I would eliminate the need for two calls before the fit. fit would operate on the preprocessed and splitter data.

on a more general note, I think it is worth asking what functionality you are adding on top of sklearn, in other words what prevents you from using sklearn all the way. two additions here seem to be preparing the data by discarding some and splitting them which could be done outside the model or in a separate class called `DataLoader` or something. other than that I see a different evaluate and a save method which justify the use of a custom model. the reason I am mentioning that is that whenever we decide to introduce custom classes we make it slightly less maintainable in the sense that we need to maintain the functionality works not the sklearn community which is fine of course when there is a reason.